### PR TITLE
Feature/configuration

### PR DESCRIPTION
**Pull request by jhetrifork**
The KDW configuration interface supporting:

- Intent configuration for receiving scanner status updates and scanned barcode results
- Barcode configuration for configuring enabled barcode formats and the individual settings for each format

### CODE REVIEW

File: kotlin-data-wedge-lib/src/main/java/dk/gls/kdw/configuration/scanner/DataWedgeHardwareScanner.kt [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+                    // Received a scanner or profile switch status notification
+                    if (intent.action == ACTION_RESULT_NOTIFICATION) {
+                        val extras = intent.getBundleExtra(EXTRA_RESULT_NOTIFICATION)
+                        val notificationType = extras?.getString(EXTRA_RESULT_NOTIFICATION_TYPE)
+                        when (notificationType) {
+                            // The scanner changed status
+                            EXTRA_KEY_VALUE_SCANNER_STATUS -> {
+                                val scannerStatus = extras.getString(EXTRA_KEY_VALUE_NOTIFICATION_STATUS)
+
+                                //TODO Add error handling in case scanner status is null?
```
Comment (esotrifork) [inline kotlin-data-wedge-lib/src/main/java/dk/gls/kdw/configuration/scanner/DataWedgeHardwareScanner.kt]: Do we need error handling and what is the behaviour if status is null?

File: kotlin-data-wedge-lib/src/test/java/dk/gls/kdw/AppConfigurationTest.kt [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import org.junit.Assert
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.RobolectricTestRunner
+
+@RunWith(RobolectricTestRunner::class)
+class AppConfigurationTest {
+
+    @Test
+    fun app_configuration_bundle_test() {
```
Comment (esotrifork) [inline kotlin-data-wedge-lib/src/test/java/dk/gls/kdw/AppConfigurationTest.kt]: I prefer the test cases named after what they do
e.g. Bundling_config_should_keep_all_values_unchanged

**Comment by jhetrifork** [comment]
@esotrifork @mnlatrifork I would love some feedback ❤️

File: kotlin-data-wedge-lib/src/main/java/dk/gls/kdw/configuration/scanner/DataWedgeHardwareScanner.kt (hunk 1 above)
Comment (jhetrifork) [reply kotlin-data-wedge-lib/src/main/java/dk/gls/kdw/configuration/scanner/DataWedgeHardwareScanner.kt]: No. I removed the TODO.

**Review by esotrifork — APPROVED** [review]

File: kotlin-data-wedge-lib/src/main/java/dk/gls/kdw/model/barcode/model/CheckDigit.kt
Comment (reivtrifork) [inline kotlin-data-wedge-lib/src/main/java/dk/gls/kdw/model/barcode/model/CheckDigit.kt:1]: Like the comments on these enum class, could this file perhaps benefit with some?

File: kotlin-data-wedge-lib/src/main/java/dk/gls/kdw/model/barcode/composite/CompositeAB.kt [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import dk.gls.kdw.model.barcode.model.UCCLinkMode
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class CompositeAB(
+    @SerialName("decoder_composite_ab")
+    val enabled: Boolean = true,
+    /**
+     * Select an option for linking UPC barcodes with a 2D barcode during transmission as if they were one barcode
```
Comment (reivtrifork) [inline kotlin-data-wedge-lib/src/main/java/dk/gls/kdw/model/barcode/composite/CompositeAB.kt:14]: Says UPC in comments but is named ucc - are they the same?

File: kotlin-data-wedge-lib/src/main/java/dk/gls/kdw/model/barcode/I2of5.kt [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    /**
+     * Sets the reader to read the barcode twice before accepting data.
+     */
+    @SerialName("decoder_i2of5_redundancy")
+    val redundancy: Boolean,
+
+    /**
+     * 	Transmits Interleaved 2 of 5 data with the check digit when enabled.
+     */
+    @SerialName("decoder_code11_report_check_digit")
```
Comment (reivtrifork) [inline kotlin-data-wedge-lib/src/main/java/dk/gls/kdw/model/barcode/I2of5.kt:37]: Is this the correct decoder serial name?

File: kotlin-data-wedge-lib/src/main/java/dk/gls/kdw/configuration/scanner/ParityFlowScannerController.kt [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+/**
+ * The parity flow scanner controller filters unnecessary calls to the underlying [DataWedgeHardwareScanner] by checking the last reported scanner status before sending new resume or suspend commands
+ * Therefore this scanner controllers [resumeScanner] and [suspendScanner] methods can safely be called multiple times without resulting in unnecessary broadcast intents being send to the Data wedge
+ */
+class ParityFlowScannerController : ScannerController() {
+
+    //We have an actual scanner status flow and a wished scanner status flow for combining later
+    private val actualScannerStatusEnumFlow = MutableSharedFlow<ScannerSimpleStatus>(replay = 1)
+    private val wishedScannerStatusFlow = MutableSharedFlow<ScannerSimpleStatus>(replay = 1)
```
Comment (reivtrifork) [inline kotlin-data-wedge-lib/src/main/java/dk/gls/kdw/configuration/scanner/ParityFlowScannerController.kt]: Super minor nitpick: wished seems like a danglish translation. Perhaps desired/expected would make more sense?

File: kotlin-data-wedge-lib/src/main/java/dk/gls/kdw/model/barcode/composite/CompositeAB.kt (hunk 3 above)
Comment (jhetrifork) [reply kotlin-data-wedge-lib/src/main/java/dk/gls/kdw/model/barcode/composite/CompositeAB.kt:14]: Actually a good question. But this is copied from the official documentation. As I have based the entire library and the documentation on that, I think we have too just trust it.

File: kotlin-data-wedge-lib/src/main/java/dk/gls/kdw/model/barcode/I2of5.kt (hunk 4 above)
Comment (jhetrifork) [reply kotlin-data-wedge-lib/src/main/java/dk/gls/kdw/model/barcode/I2of5.kt:37]: Seems like a weird name, but yes according to the documentation.

**Pull request merged**